Update doc with gRPC JSON transcoding (#25825)
dotnet/AspNetCore.Docs: aspnetcore/grpc/httpapi.md

### Usage

> 1. Add a package reference to [Microsoft.AspNetCore.Grpc.JsonTranscoding](https://www.nuget.org/packages/Microsoft.AspNetCore.Grpc.JsonTranscoding).

Reviewer: ```suggestion
1. Add a package reference to [`Microsoft.AspNetCore.Grpc.JsonTranscoding`](https://www.nuget.org/packages/Microsoft.AspNetCore.Grpc.JsonTranscoding).
```

> 1. Register it in server startup code by adding `AddJsonTranscoding`. For example, `services.AddGrpc().AddJsonTranscoding()`.

Reviewer: ```suggestion
1. Register the package in server startup code by adding `AddJsonTranscoding`. For example, `services.AddGrpc().AddJsonTranscoding()`.
```

> 1. Add [google/api/http.proto](https://github.com/dotnet/aspnetcore/blob/8b601c3a73ba66de4e6ca35530b5d32a48c76c5b/src/Grpc/JsonTranscoding/test/testassets/Sandbox/google/api/http.proto) and [google/api/annotations.proto](https://github.com/dotnet/aspnetcore/blob/main/src/Grpc/JsonTranscoding/test/testassets/Sandbox/google/api/annotations.proto) files to your project.

Reviewer: ```suggestion
1. Add [`google/api/http.proto`](https://github.com/dotnet/aspnetcore/blob/8b601c3a73ba66de4e6ca35530b5d32a48c76c5b/src/Grpc/JsonTranscoding/test/testassets/Sandbox/google/api/http.proto) and [`google/api/annotations.proto`](https://github.com/dotnet/aspnetcore/blob/main/src/Grpc/JsonTranscoding/test/testassets/Sandbox/google/api/annotations.proto) files to the project.
```

1. Annotate gRPC methods in your `.proto` files with HTTP bindings and routes:

[!code-protobuf[](~/grpc/httpapi/greet.proto?highlight=3,9-11)]

The `SayHello` gRPC method can now be invoked as gRPC and as a JSON Web API:

* Request: `HTTP/1.1 GET /v1/greeter/world`
* Response: `{ "message": "Hello world" }`

> If the server is configured to write logs for each request, server logs show that the HTTP call is executed by a gRPC service. JSON transcoding maps the incoming HTTP request to a gRPC message, and then converts the response message to JSON.

Reviewer: ```suggestion
If the server is configured to write logs for each request, server logs show that the HTTP call is executed by a gRPC service. JSON transcoding maps the incoming HTTP request to a gRPC message and then converts the response message to JSON.
```

This is a basic example. See [HttpRule](https://cloud.google.com/service-infrastructure/docs/service-management/reference/rpc/google.api#google.api.HttpRule) for more customization options.

### OpenAPI support

> JSON transcoding currently doesn't support OpenAPI. During the .NET 7 development process the .NET team will investigate the best way to support OpenAPI.

Reviewer: ```suggestion
JSON transcoding currently doesn't support OpenAPI. During the development of .NET 7, the .NET team will investigate the best way to support OpenAPI.
```

### JSON transcoding vs gRPC-Web

Both JSON transcoding and gRPC-Web allow gRPC services to be called from a browser. However, the way each does this is different:

* gRPC-Web lets browser apps call gRPC services from the browser with the gRPC-Web client and Protobuf. gRPC-Web requires the browser app generate a gRPC client, and has the advantage of sending small, fast Protobuf messages.
* JSON transcoding allows browser apps to call gRPC services as if they were RESTful APIs with JSON. The browser app doesn't need to generate a gRPC client or know anything about gRPC.

No generated client is created for JSON transcoding. The previous `Greeter` service can be called using browser JavaScript APIs:

```javascript
var name = nameInput.value;

fetch('/v1/greeter/' + name)

>     .then((response) => response.json())
>     .then((result) => {
>         console.log(result.message);
>         // Hello world
>     });

Reviewer: ```suggestion
  .then((response) => response.json())
  .then((result) => {
    console.log(result.message);
    // Hello world
  });
```

```

## grpc-gateway

[grpc-gateway](https://grpc-ecosystem.github.io/grpc-gateway/) is another technology for creating RESTful JSON APIs from gRPC services. It uses the same `.proto` annotations to map HTTP concepts to gRPC services.

> grpc-gateway uses code generation to create a reverse-proxy server. The reverse-proxy translates RESTful calls into gRPC+Protobuf and sends them over HTTP/2 to the gRPC service. The benefit of this approach is the gRPC service doesn't know about the RESTful JSON APIs. Any gRPC server can use grpc-gateway.

Reviewer: ```suggestion
grpc-gateway uses code generation to create a reverse-proxy server. The reverse proxy translates RESTful calls into gRPC+Protobuf and sends the calls over HTTP/2 to the gRPC service. The benefit of this approach is the gRPC service doesn't know about the RESTful JSON APIs. Any gRPC server can use grpc-gateway.
```

Meanwhile, gRPC JSON transcoding runs inside an ASP.NET Core app. It deserializes JSON into Protobuf messages, then invokes the gRPC service directly. JSON transcoding in ASP.NET Core offers advantages to .NET app developers:

> * Fewer moving parts. Both gRPC services and mapped RESTful JSON API run out of one ASP.NET Core application.

Reviewer: ```suggestion
* Less complex: Both gRPC services and mapped RESTful JSON API run out of one ASP.NET Core app.
```

> * Performance. JSON transcoding deserializes JSON to Protobuf messages and invokes the gRPC service directly. There are significant performance benefits in doing this in-process vs. making a new gRPC call to a different server.

Reviewer: ```suggestion
* Better performance: JSON transcoding deserializes JSON to Protobuf messages and invokes the gRPC service directly. There are significant performance benefits in doing this in-process versus making a new gRPC call to a different server.
```

> * Cost. Fewer servers = smaller monthly hosting bill.

Reviewer: ```suggestion
* Lower cost: Fewer servers result in a smaller monthly hosting bill.
```
